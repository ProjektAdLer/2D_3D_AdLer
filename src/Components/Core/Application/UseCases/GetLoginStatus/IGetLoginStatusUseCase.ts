import { ISynchronousUsecase } from "../../Abstract/ISynchronousUsecase";
import LoginStatusTO from "../../DataTransferObjects/LoginStatusTO";

export default interface IGetLoginStatusUseCase
  extends ISynchronousUsecase<void, LoginStatusTO> {}
