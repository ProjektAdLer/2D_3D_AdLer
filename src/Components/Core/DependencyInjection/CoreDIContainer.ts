import { LoadWorldController } from "../Presentation/React/LoadWorldButton/LoadWorldController";
import ILoadWorldController from "../Presentation/React/LoadWorldButton/ILoadWorldController";
import { Container } from "inversify";
import CORE_TYPES from "./CoreTypes";
import ISceneView from "../Presentation/Babylon/SceneManagement/ISceneView";
import SceneView from "../Presentation/Babylon/SceneManagement/SceneView";
import SceneViewModel from "../Presentation/Babylon/SceneManagement/SceneViewModel";
import ScenePresenter from "../Presentation/Babylon/SceneManagement/ScenePresenter";
import MainScene from "../Presentation/Babylon/SceneManagement/MainScene";
import ICreateSceneClass from "../Presentation/Babylon/SceneManagement/ICreateSceneClass";
import IScenePresenter from "../Presentation/Babylon/SceneManagement/IScenePresenter";
import infrastructureDIContainer from "./InfrastructureDIContainer";
import useCaseDIContainer from "./UseCases/UseCaseDIConatiner";
import PortsDIContainer from "./Ports/PortsDIContainer";
import BuilderDIContainer from "./Builders/BuilderDIContainer";
import INavigation from "../Presentation/Babylon/Navigation/INavigation";
import Navigation from "../Presentation/Babylon/Navigation/Navigation";
import NavigationConfiguration from "../Presentation/Babylon/Navigation/NavigationConfiguration";

var CoreDIContainer = new Container();

// Scene
CoreDIContainer.bind<ISceneView>(CORE_TYPES.ISceneView)
  .to(SceneView)
  .inSingletonScope();
CoreDIContainer.bind(SceneViewModel).toSelf().inSingletonScope();
CoreDIContainer.bind<IScenePresenter>(CORE_TYPES.IScenePresenter)
  .to(ScenePresenter)
  .inSingletonScope();

// Navigation
CoreDIContainer.bind<INavigation>(CORE_TYPES.INavigation)
  .to(Navigation)
  .inSingletonScope();
CoreDIContainer.bind(NavigationConfiguration).toSelf();

// Controllers
CoreDIContainer.bind<ILoadWorldController>(CORE_TYPES.ILoadWorldController)
  .to(LoadWorldController)
  .inSingletonScope();

// bind other CreateSceneClass here for testing puposes -MK
CoreDIContainer.bind<ICreateSceneClass>(CORE_TYPES.ICreateSceneClass).to(
  MainScene
);

CoreDIContainer.load(infrastructureDIContainer);
CoreDIContainer.load(useCaseDIContainer);
CoreDIContainer.load(PortsDIContainer);
CoreDIContainer.load(BuilderDIContainer);

export default CoreDIContainer;
