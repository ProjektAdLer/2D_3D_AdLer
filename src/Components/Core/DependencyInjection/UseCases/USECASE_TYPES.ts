const USECASE_TYPES = {
  ILoadUserLearningWorldsUseCase: Symbol("ILoadUserLearningWorldsUseCase"),
  ILoadLearningWorldUseCase: Symbol("ILoadLearningWorldUseCase"),
  ILoadAvatarUseCase: Symbol("ILoadAvatarUseCase"),
  ILoadLearningElementUseCase: Symbol("ILoadLearningElementUseCase"),
  IScoreLearningElementUseCase: Symbol("IScoreLearningElementUseCase"),
  ICalculateLearningSpaceScoreUseCase: Symbol(
    "ICalculateLearningSpaceScoreUseCase"
  ),
  ICalculateLearningWorldScoreUseCase: Symbol(
    "ICalculateLearningWorldScoreUseCase"
  ),
  ILoginUseCase: Symbol("ILoginUseCase"),
  ILoadLearningSpaceUseCase: Symbol("ILoadLearningSpaceUseCase"),
  IScoreH5PLearningElementUseCase: Symbol("IScoreH5PLearningElementUseCase"),
  IGetLearningElementSourceUseCase: Symbol("IGetLearningElementSourceUseCase"),
  IGetLoginStatusUseCase: Symbol("IGetLoginStatusUseCase"),
  ISetUserLocationUseCase: Symbol("ISetUserLocationUseCase"),
  IGetUserLocationUseCase: Symbol("IGetUserLocationUseCase"),
  ICalculateLearningSpaceAvailabilityUseCase: Symbol(
    "ICalculateLearningSpaceAvailabilityUseCase"
  ),
  IGetLearningSpacePrecursorAndSuccessorUseCase: Symbol(
    "IGetLearningSpacePrecursorAndSuccessorUseCase"
  ),
  ILoadAdaptivityElementUseCase: Symbol("ILoadAdaptivityElementUseCase"),
  IGetAdaptivityElementStatusUseCase: Symbol(
    "IGetAdaptivityElementStatusUseCase"
  ),
  ISubmitAdaptivityElementSelectionUseCase: Symbol(
    "ISubmitAdaptivityElementSelectionUseCase"
  ),
  IStartOverallTimeSpentNotificationTimerUseCase: Symbol(
    "IStartOverallTimeSpentNotificationTimerUseCase"
  ),
  ICreateOverallTimeSpentNotificationTimerUseCase: Symbol(
    "ICreateOverallTimeSpentNotificationTimerUseCase"
  ),
  IPauseOverallTimeSpentNotificationTimerUseCase: Symbol(
    "IPauseOverallTimeSpentNotificationTimerUseCase"
  ),
};

export default USECASE_TYPES;
