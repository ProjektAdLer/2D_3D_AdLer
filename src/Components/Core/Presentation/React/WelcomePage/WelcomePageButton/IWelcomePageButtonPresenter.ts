import ILMSAdapter from "src/Components/Core/Application/Ports/LMSPort/ILMSAdapter";

export default interface IButtonPresenter extends ILMSAdapter {}
