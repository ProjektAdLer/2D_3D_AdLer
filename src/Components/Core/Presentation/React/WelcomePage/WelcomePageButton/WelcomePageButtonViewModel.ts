import Observable from "../../../../../../Lib/Observable";

export default class WelcomePageButtonViewModel {
  userLoggedIn = new Observable<boolean>();
}
