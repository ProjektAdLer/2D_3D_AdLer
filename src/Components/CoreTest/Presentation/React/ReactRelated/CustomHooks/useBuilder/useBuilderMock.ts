import * as useBuilder from "~ReactComponents/ReactRelated/CustomHooks/useBuilder";

/**
 * Mocks the useViewModelControllerProvider hook.
 * @param mockedTupels The return value, which the mocked hook will return.
 *
 */
export default function useBuilderMock(
  mockedTupels: [viewModel: any, controller: any] = [undefined, undefined]
) {
  jest.spyOn(useBuilder, "default").mockReturnValue(mockedTupels);
}
