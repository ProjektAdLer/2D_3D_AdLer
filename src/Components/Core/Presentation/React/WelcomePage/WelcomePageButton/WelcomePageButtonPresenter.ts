import IWelcomePageButtonPresenter from "./IWelcomePageButtonPresenter";
import WelcomePageButtonViewModel from "./WelcomePageButtonViewModel";

export default class WelcomePageButtonPresenter
  implements IWelcomePageButtonPresenter
{
  constructor(private viewModel: WelcomePageButtonViewModel) {}

  onLoginSuccessful(): void {
    this.viewModel.userLoggedIn.Value = true;
  }

  onLogoutSuccessful(): void {
    this.viewModel.userLoggedIn.Value = false;
  }
}
