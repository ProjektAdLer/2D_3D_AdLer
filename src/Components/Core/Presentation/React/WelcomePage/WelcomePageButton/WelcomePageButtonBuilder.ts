import { injectable } from "inversify";
import PresentationBuilder from "../../../PresentationBuilder/PresentationBuilder";
import CoreDIContainer from "../../../../DependencyInjection/CoreDIContainer";
import PORT_TYPES from "../../../../DependencyInjection/Ports/PORT_TYPES";
import WelcomePageButtonViewModel from "./WelcomePageButtonViewModel";
import WelcomePageButtonPresenter from "./WelcomePageButtonPresenter";
import AbstractPort from "src/Components/Core/Application/Ports/AbstractPort/AbstractPort";
import ILMSAdapter from "src/Components/Core/Application/Ports/LMSPort/ILMSAdapter";
import { HistoryWrapper } from "~ReactComponents/ReactRelated/ReactEntryPoint/HistoryWrapper";

@injectable()
export default class WelcomePageButtonBuilder extends PresentationBuilder<
  WelcomePageButtonViewModel,
  undefined,
  undefined,
  WelcomePageButtonPresenter
> {
  constructor() {
    super(
      WelcomePageButtonViewModel,
      undefined,
      undefined,
      WelcomePageButtonPresenter,
    );
  }

  override buildPresenter(): void {
    super.buildPresenter();
    CoreDIContainer.get<AbstractPort<ILMSAdapter>>(
      PORT_TYPES.ILMSPort,
    ).registerAdapter(this.presenter!, HistoryWrapper.currentLocationScope());
  }
}
