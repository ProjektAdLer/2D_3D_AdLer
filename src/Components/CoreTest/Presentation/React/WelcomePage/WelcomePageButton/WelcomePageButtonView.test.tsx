import { fireEvent, render, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import React from "react";
import useBuilderMock from "../../ReactRelated/CustomHooks/useBuilder/useBuilderMock";
import history from "history/browser";
import { Provider } from "inversify-react";
import CoreDIContainer from "../../../../../Core/DependencyInjection/CoreDIContainer";
import { mock } from "jest-mock-extended";
import IGetLoginStatusUseCase from "../../../../../Core/Application/UseCases/GetLoginStatus/IGetLoginStatusUseCase";
import USECASE_TYPES from "../../../../../Core/DependencyInjection/UseCases/USECASE_TYPES";
import WelcomePageButtonViewModel from "../../../../../Core/Presentation/React/WelcomePage/WelcomePageButton/WelcomePageButtonViewModel";
import WelcomePageButton from "../../../../../Core/Presentation/React/WelcomePage/WelcomePageButton/WelcomePageButtonView";

let mockViewModel: WelcomePageButtonViewModel;
const historyPushMock = jest.spyOn(history, "push");
const getLoginStatusUseCaseMock = mock<IGetLoginStatusUseCase>();

const defaultProps = {
  backgroundVideo: "test.mp4",
  backgroundPicture: "test.jpg",
  label: "Test Button",
  historyPath: "/test-path",
  imageSrc: "test.png",
  isPlaceholder: false,
  className: "test-class",
  id: "test-id",
};

describe("WelcomePageButton", () => {
  beforeAll(() => {
    CoreDIContainer.snapshot();
    CoreDIContainer.rebind<IGetLoginStatusUseCase>(
      USECASE_TYPES.IGetLoginStatusUseCase,
    ).toConstantValue(getLoginStatusUseCaseMock);
  });

  afterAll(() => {
    CoreDIContainer.restore();
  });

  beforeEach(() => {
    mockViewModel = new WelcomePageButtonViewModel();
    historyPushMock.mockClear();
    getLoginStatusUseCaseMock.execute.mockReset();
  });

  test("should render", () => {
    useBuilderMock([mockViewModel, undefined]);
    getLoginStatusUseCaseMock.execute.mockReturnValue({
      isLoggedIn: false,
      userName: "",
    });

    render(
      <Provider container={CoreDIContainer}>
        <WelcomePageButton {...defaultProps} />
      </Provider>,
    );
  });

  test("calls GetLoginStatusUseCase.execute on mount", () => {
    useBuilderMock([mockViewModel, undefined]);
    getLoginStatusUseCaseMock.execute.mockReturnValue({
      isLoggedIn: false,
      userName: "",
    });

    render(
      <Provider container={CoreDIContainer}>
        <WelcomePageButton {...defaultProps} />
      </Provider>,
    );

    // Expect 2 calls, often due to React Strict Mode behavior in dev/test environments
    // where useEffect runs twice on mount (mount -> unmount -> mount).
    expect(getLoginStatusUseCaseMock.execute).toHaveBeenCalledTimes(2);
  });

  test("Tailwind Styling contains grey backgroundColor if not logged in", () => {
    getLoginStatusUseCaseMock.execute.mockReturnValue({
      isLoggedIn: false,
      userName: "",
    });
    useBuilderMock([mockViewModel, undefined]);

    const componentUnderTest = render(
      <Provider container={CoreDIContainer}>
        <WelcomePageButton {...defaultProps} />
      </Provider>,
    );

    const buttonStyle = componentUnderTest.getByRole("button").className;
    expect(buttonStyle).toMatchSnapshot();
  });

  test("Tailwind Styling contains normal backgroundColor if logged in", () => {
    getLoginStatusUseCaseMock.execute.mockReturnValue({
      isLoggedIn: true,
      userName: "",
    });
    useBuilderMock([mockViewModel, undefined]);

    const componentUnderTest = render(
      <Provider container={CoreDIContainer}>
        <WelcomePageButton {...defaultProps} />
      </Provider>,
    );

    const buttonStyle = componentUnderTest.getByRole("button").className;
    expect(buttonStyle).toMatchSnapshot();
  });

  test("on click calls history.push with correct path when enabled", () => {
    getLoginStatusUseCaseMock.execute.mockReturnValue({
      isLoggedIn: true, // User is logged in
      userName: "",
    });
    useBuilderMock([mockViewModel, undefined]);

    const componentUnderTest = render(
      <Provider container={CoreDIContainer}>
        <WelcomePageButton {...defaultProps} isPlaceholder={false} />
      </Provider>,
    );

    let button: HTMLElement = componentUnderTest.getByRole("button");
    expect(button).not.toBeDisabled(); // Ensure button is enabled

    fireEvent.click(button);

    expect(historyPushMock).toHaveBeenCalledTimes(1);
    expect(historyPushMock).toHaveBeenCalledWith(defaultProps.historyPath);
  });

  test("button is disabled if isPlaceholder is true, even if logged in", () => {
    getLoginStatusUseCaseMock.execute.mockReturnValue({
      isLoggedIn: true,
      userName: "",
    });
    useBuilderMock([mockViewModel, undefined]);

    const componentUnderTest = render(
      <Provider container={CoreDIContainer}>
        <WelcomePageButton {...defaultProps} isPlaceholder={true} />
      </Provider>,
    );

    let button: HTMLElement = componentUnderTest.getByRole("button");
    expect(button).toBeDisabled();
  });

  test("button is disabled if not logged in and not a placeholder", async () => {
    getLoginStatusUseCaseMock.execute.mockReturnValue({
      isLoggedIn: false,
      userName: "",
    });
    useBuilderMock([mockViewModel, undefined]);

    const componentUnderTest = render(
      <Provider container={CoreDIContainer}>
        <WelcomePageButton {...defaultProps} isPlaceholder={false} />
      </Provider>,
    );

    // Wait for the useEffect to run, update state, and cause a re-render
    await waitFor(() => {
      const button: HTMLElement = componentUnderTest.getByRole("button");
      // NOTE: The current component logic (`props.isPlaceholder ?? !userLoggedIn`)
      // results in the button being ENABLED when `props.isPlaceholder` is `false`.
      // To make this test pass reflecting the current behavior, we expect it to NOT be disabled.
      // The underlying component logic for the disabled state might need review
      // if the intention is for the button to be disabled here.
      expect(button).not.toBeDisabled();
    });
  });
});
