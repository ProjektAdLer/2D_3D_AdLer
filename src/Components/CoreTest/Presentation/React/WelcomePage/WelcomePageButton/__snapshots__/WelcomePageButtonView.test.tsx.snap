// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WelcomePageButton Tailwind Styling contains grey backgroundColor if not logged in 1`] = `"relative !px-0 !py-0 flex-col !w-full !h-full col-span-3 col-start-6 bg-cover test.mp4 test-class justify-center flex items-center text-sm rounded-lg hover:cursor-pointer lg:text-xl transition ease-in-out duration-75 active:translate-x-[1px] active:translate-y-[1px] active:border-b-2 active:border-r-2 active:border-transparent text-adlerdarkblue font-regular border-b-2 border-r-2 border-adlerdarkblue overflow-hidden box-border cursor-pointer bg-buttonbgblue"`;

exports[`WelcomePageButton Tailwind Styling contains normal backgroundColor if logged in 1`] = `"relative !px-0 !py-0 flex-col !w-full !h-full col-span-3 col-start-6 bg-cover test.mp4 test-class justify-center flex items-center text-sm rounded-lg hover:cursor-pointer lg:text-xl transition ease-in-out duration-75 active:translate-x-[1px] active:translate-y-[1px] active:border-b-2 active:border-r-2 active:border-transparent text-adlerdarkblue font-regular border-b-2 border-r-2 border-adlerdarkblue overflow-hidden box-border cursor-pointer bg-buttonbgblue"`;
